src/Version.mli

```diff
@@ -0,0 +1,22 @@
+(**********************************************************************
+ *                                                                    *
+ *                            OCamlFormat                             *
+ *                                                                    *
+ *  Copyright (c) 2017-present, Facebook, Inc.  All rights reserved.  *
+ *                                                                    *
+ *  This source code is licensed under the MIT license found in the   *
+ *  LICENSE file in the root directory of this source tree.           *
+ *                                                                    *
+ **********************************************************************)
+
+val version : string
+(** A version number, or "unknown". This is provided by [dune-build-info],
```

Comment:
This string will be compared with the `version` field from .ocamlformat files and the error message will be
`some_obfuscated_message ^ "expecting unknown but got 0.11.0"`.

Should this be an `option` instead ? So the doc can say "unknown" while the version check is disabled.

Reply:
I thought a legit version number was set when building, if not is there a way to do it?
Anyway I think it would be nice to have a type to convey more information like: `Released of int * int * int | Unreleased of ...`

Reply:
> Anyway I think it would be nice to have a type to convey more information like: `Released of int * int * int | Unreleased of ...`

Yes, we need at least that to relax the version check with dev builds (see https://github.com/ocaml-ppx/ocamlformat/pull/1147).

Reply:
Note that not performing version check on dev builds would break our deployment. Wouldn't it be possible to use a command line flag or environment variable instead?

Reply:
Yes, that's something we can improve. But I think that this problem already exists in `master`, right? `dune exec ocamlformat -- --version` will return something like `0.12.0-00000-dirty` and try to compare that with the version in the config file.

Reply:
Yes, and that it fails in that situation is crucial. (Well, I'm a couple days behind master.)

Reply:
@jberdine there are two different meanings of "dev builds" here - binaries under `_build` and installed binaries from an unreleased version. The latter would have an actual version that is set.

Reply:
> Note that not performing version check on dev builds would break our deployment. Wouldn't it be possible to use a command line flag or environment variable instead?

Are you using version strings of the form `0.12-25-g6bdc881` in your `.ocamlformat` files ?

To support that, we could match the version depending on the precision given by the user:
If we use version strings of the form `a.b.c-commit`:
- `0.12` matches any `0.12.*-*`
- `0.12.0` matches only `0.12.0-*`
- `0.12-hash` match precisely one commit: `0.12.*-hash`

Reply:
yes, that behavior would make sense.

Reply:
let's continue that in #1363 since it's mostly independent from using `dune-build-info` vs `gen_version.mlt`.